### PR TITLE
Fix clock connections in InferReadWrite

### LINKED ISSUE #307: Correct clock connections in InferReadWrite.
Clocks signals should be correctly connected by examining ports' clocks.
(found in #306)

### PR DESCRIPTION
**Pull request by donggyukim**
* SHOULD BE MERGED AFTER #306 AND CHANGE BASE TO MASTER

This fixes #307, and should be merged quickly due to ucb-bar/chisel3#288.

### CODE REVIEW

**Comment by aswaterman** [comment]
Thanks @donggyukim, this works for me

File: src/main/scala/firrtl/passes/AnnotateMemMacros.scala [hunk 1]
```diff
@@ -72,6 +72,8 @@ object AnalysisUtils {
       if (nodeWidth == extractionWidth) getOrigin(connects, args.head) else e
     case DoPrim((PrimOps.AsUInt | PrimOps.AsSInt | PrimOps.AsClock), args, _, _) => 
       getOrigin(connects, args.head)
+    // Todo: It's not clear it's ok to call remove validifs before mem passes...
+    case ValidIf(cond, value, ClockType) => getOrigin(connects, value)
```
Comment (shunshou) [inline src/main/scala/firrtl/passes/AnnotateMemMacros.scala]: I thought @jackkoenig said ValidIf wouldn't be in this level of FIRRTL?

File: src/test/scala/firrtlTests/InferReadWriteSpec.scala [hunk 2]
```diff
@@ -103,4 +106,36 @@ circuit sram6t :
...
+      T_2 <= io.raddr
+    read mport T_3 = mem[T_2], clk1
+    io.rdata <= T_3
+    node T_4 = and(io.en, io.wen)
+    when T_4 :
+      write mport T_5 = mem[io.waddr], clk2
+      T_5 <= io.wdata
+""".stripMargin
+
+    val annotaitonMap = AnnotationMap(Seq(InferReadWriteAnnotation("sram6t", TransID(-1))))
```
Comment (shunshou) [inline src/test/scala/firrtlTests/InferReadWriteSpec.scala]: Also, "annotationMap"

Comment (donggyukim) [inline src/test/scala/firrtlTests/InferReadWriteSpec.scala]: Thanks, I made this typo a lot

**Comment by aswaterman** [comment]
is this safe to merge?

**Comment by donggyukim** [comment]
@aswaterman Yes.

File: src/main/scala/firrtl/passes/AnnotateMemMacros.scala (hunk 1 above)
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/AnnotateMemMacros.scala]: There's nothing wrong with ValidIf existing in low Firrtl, eventually we might even use it in the emitter

**Comment by aswaterman** [comment]
thanks, i'll merge once tests pass

**Pull request merged**